app/models/hint.rb

```diff
@@ -1,5 +1,36 @@
+require 'stringex/core_ext'
+
 class Hint < ApplicationRecord
   def self.match(searchterm)
-    Hint.find_by(fingerprint: searchterm)
+    searchprint = fingerprint(searchterm)
+    Hint.find_by(fingerprint: searchprint)
   end
+
+  # This implements the OpenRefine fingerprinting algorithm. See
+  # https://github.com/OpenRefine/OpenRefine/wiki/Clustering-In-Depth#fingerprint .
+  def self.fingerprint(searchterm)
+    # Preprocess.
+    searchterm.strip!
+    searchterm.downcase!
+    # What's up with the weird regexp? We want to remove punctuation from
+    # the string EXCEPT where it where it is part of the term C++ or C#
+    # (as a standalone term - appearing either immediately after whitespace
+    # or at the beginning of a line). Those terms are meaningfully different
+    # from C, and our users frequently search for them.
+    searchterm.gsub!(/(?<!\s[cC]|\s[cC]\+|^[cC]|^[cC]\+)[[:punct:]]/, '')
+    # Tokenize.
+    term_array = searchterm.split(' ')
+
+    # Remove duplicates and sort.
+    term_array.uniq!
+    term_array.sort!
+
+    # Rejoin tokens.
+    fingerprint = term_array.join(' ')
+
+    # Normalize to ASCII (e.g. gödel and godel are liable to be intended to
+    # find the same thing). (Yes, to_ascii lacks a to_ascii! equivalent.)
+    fingerprint.to_ascii
+  end
+
```

Comment:
Naturally this is a thing that is _preferred_ for Python, and I had therefore configured my editor to auto-add newlines at EOF. *figures out how to make that setting language-specific*